### PR TITLE
BUGFIX: SD-921 Sorting legend items by alphabet order

### PR DESCRIPTION
**Pull request by dlcd2203**
---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [ ] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [ ] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [ ] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [ ] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [ ] Successful `extended test - examples`
- [ ] Successful `extended test - storybook`
- [ ] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)

# Related PRs

- gdc-analytical-designer: https://github.com/gooddata/gdc-analytical-designer/pull/3032
- gdc-dashboards: https://github.com/gooddata/gdc-dashboards/pull/2748

# Related Jira tasks

SD-921: https://jira.intgdc.com/browse/SD-921

### CODE REVIEW

**Comment by dlcd2203** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7077/ : SUCCESS in 8m 32s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3026/ : SUCCESS in 8s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3703/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3027/ : SUCCESS in 15s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/3072/ : UNSTABLE in 19m 22s

**Comment by dlcd2203** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7081/ : SUCCESS in 5m 30s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3034/ : SUCCESS in 6s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3705/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3035/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/3076/ : UNSTABLE in 29m 58s

**Comment by dlcd2203** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7097/ : SUCCESS in 10m 43s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3058/ : SUCCESS in 3s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3711/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3059/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/3088/ : SUCCESS in 19m 04s

**Comment by dlcd2203** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1552.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3060/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/3089/ : SUCCESS in 20m 38s

File: src/helpers/sorts.ts [hunk 1]
```diff
@@ -103,6 +103,16 @@ export function getFirstMeasureSort(afm: AFM.IAfm): AFM.SortItem[] {
     return [];
 }
 
+export function getDefaultAttributeGeoPushpinSort(afm: AFM.IAfm): AFM.SortItem[] {
+    const attributes: AFM.IAttribute[] = get(afm, "attributes");
+
+    if (attributes.length === 2 || attributes.length === 3) {
```
Comment (ptquang86) [inline src/helpers/sorts.ts:109]: how about `if (attributes.length > 1)`? 
Please add some comment here about this sort

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7098/ : SUCCESS in 9m 14s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3061/ : SUCCESS in 3s

File: src/helpers/sorts.ts [hunk 2]
```diff
@@ -103,6 +103,16 @@ export function getFirstMeasureSort(afm: AFM.IAfm): AFM.SortItem[] {
...
+        return getAttributeSortItems(attributes[1].localIdentifier, ASC);
```
Comment (ptquang86) [inline src/helpers/sorts.ts:110]: what if `attributes[1]` is tooltipText, not segmentBy?

File: src/internal/components/pluggableVisualizations/geoChart/PluggableGeoPushpinChart.tsx [hunk 3]
```diff
@@ -259,12 +259,14 @@ export class PluggableGeoPushpinChart extends PluggableBaseChart {
             dimensions: this.getDimensions(mdObject),
         };
 
-        const hasSizeMeasure = mdObject.buckets.some(bucket => bucket.localIdentifier === BucketNames.SIZE);
+        const hasSegmentAttribute = mdObject.buckets.some(
+            bucket => bucket.localIdentifier === BucketNames.SEGMENT,
```
Comment (ptquang86) [inline src/internal/components/pluggableVisualizations/geoChart/PluggableGeoPushpinChart.tsx:263]: please add type and return type

File: src/internal/utils/tests/sort.spec.ts [hunk 4]
```diff
@@ -379,8 +379,32 @@ describe("createSorts", () => {
...
+            {
+                attributeSortItem: {
+                    attributeIdentifier: "a2",
+                    direction: "asc",
+                },
+            },
+        ]);
+    });
+
+    it("should return emptySort if count of attributes diffirent 2 or 3", () => {
```
Comment (ptquang86) [inline src/internal/utils/tests/sort.spec.ts:393]: `empty sort ... different ...`
How about `should return empty sort if geo Pushpin chart has no segmentBy attribute`

Comment (ptquang86) [reply src/internal/utils/tests/sort.spec.ts:393]: This test should be moved to `helpers/tests/sorts.spec.ts`

Comment (dlcd2203) [reply src/internal/utils/tests/sort.spec.ts]: move to `helpers/tests/sorts.spec.ts`

File: src/internal/components/pluggableVisualizations/geoChart/PluggableGeoPushpinChart.tsx (hunk 3 above)
Comment (dlcd2203) [reply src/internal/components/pluggableVisualizations/geoChart/PluggableGeoPushpinChart.tsx:263]: fixed

File: src/helpers/sorts.ts (hunk 2 above)
Comment (dlcd2203) [reply src/helpers/sorts.ts:110]: we have checked condition at [here](https://github.com/gooddata/gooddata-react-components/blob/d5f702262048b60f0aa63b94fa4530af63e4cd86/src/internal/components/pluggableVisualizations/geoChart/PluggableGeoPushpinChart.tsx#L269), so attributes always have segmentBy

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: src/helpers/sorts.ts (hunk 1 above)
Comment (dlcd2203) [reply src/helpers/sorts.ts]: fixed

File: src/helpers/sorts.ts (hunk 2 above)
Comment (ptquang86) [reply src/helpers/sorts.ts:111]: Who know if in the future, somebody will call this function without checking that condition :)
Please move that condition into this function

File: src/helpers/sorts.ts [hunk 5]
```diff
@@ -106,7 +106,8 @@ export function getFirstMeasureSort(afm: AFM.IAfm): AFM.SortItem[] {
 export function getDefaultAttributeGeoPushpinSort(afm: AFM.IAfm): AFM.SortItem[] {
     const attributes: AFM.IAttribute[] = get(afm, "attributes");
 
-    if (attributes.length === 2 || attributes.length === 3) {
+    // sort by second attribute (Segmentby)
```
Comment (ptquang86) [inline src/helpers/sorts.ts:109]: `... segmentBy ..`

File: src/internal/components/pluggableVisualizations/geoChart/PluggableGeoPushpinChart.tsx (hunk 3 above)
Comment (ptquang86) [reply src/internal/components/pluggableVisualizations/geoChart/PluggableGeoPushpinChart.tsx:263]: I meant 
`(bucket: ABC): boolean => bucket.localIdentifier === BucketNames.SEGMENT `

File: src/internal/components/pluggableVisualizations/geoChart/PluggableGeoPushpinChart.tsx [hunk 6]
```diff
@@ -259,13 +259,13 @@ export class PluggableGeoPushpinChart extends PluggableBaseChart {
...
 
-        const hasSegmentAttribute = mdObject.buckets.some(
+        const hasSegmentAttribute: boolean = mdObject.buckets.some(
             bucket => bucket.localIdentifier === BucketNames.SEGMENT,
         );
 
         const allProperties: IVisualizationProperties = get(visualizationProperties, "properties", {});
 
-        // sort items SegmentBy alphabet order
+        // sort items SegmentBy by alphabet order
```
Comment (ptquang86) [inline src/internal/components/pluggableVisualizations/geoChart/PluggableGeoPushpinChart.tsx:268]: `sort items by segmentBy by alphabetical order`

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7104/ : SUCCESS in 6m 34s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3068/ : SUCCESS in 9s

File: src/internal/utils/tests/sort.spec.ts [hunk 7]
```diff
@@ -366,7 +366,7 @@ describe("createSorts", () => {
         expect(createSorts("table", emptyAfm, emptyResultSpec, visualizationProperties)).toEqual([]);
     });
 
-    it("should sort by segmentBy attribute if chart is geo Pushpin", () => {
+    it("should sort by second attribute if chart is geo Pushpin", () => {
```
Comment (ptquang86) [inline src/internal/utils/tests/sort.spec.ts:369]: I prefer `segmentBy attribute` than `second attribute `

File: src/helpers/tests/sorts.spec.ts [hunk 8]
```diff
@@ -121,4 +126,21 @@ describe("sorts", () => {
             expect(getFirstAttributeIdentifier(stackedResultSpec, 2)).toEqual(null);
         });
     });
+
+    describe("getDefaultAttributeGeoPushpinSort", () => {
+        it("should return second attribute if count of attributes greater than 1", () => {
```
Comment (ptquang86) [inline src/helpers/tests/sorts.spec.ts:131]: how about 
`should sort by segmentBy attribute`?

File: src/helpers/tests/sorts.spec.ts [hunk 9]
```diff
@@ -121,4 +126,21 @@ describe("sorts", () => {
...
+                {
+                    attributeSortItem: {
+                        attributeIdentifier: "a2",
+                        direction: "asc",
+                    },
+                },
+            ]);
+        });
+
+        it("should return empty sort if count of attributes less than 2", () => {
```
Comment (ptquang86) [inline src/helpers/tests/sorts.spec.ts:142]: how about 
`should not sort if segmentBy attribute is not existing `

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7110/ : SUCCESS in 3m 51s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3077/ : SUCCESS in 7s

**Comment by ptquang86** [comment]
LGTM

File: src/helpers/sorts.ts [hunk 10]
```diff
@@ -103,6 +103,17 @@ export function getFirstMeasureSort(afm: AFM.IAfm): AFM.SortItem[] {
     return [];
 }
 
+export function getDefaultAttributeGeoPushpinSort(afm: AFM.IAfm): AFM.SortItem[] {
+    const attributes: AFM.IAttribute[] = get(afm, "attributes");
```
Comment (naibinh) [inline src/helpers/sorts.ts:107]: it could be just ES6 destructuring assignment so that Ts can do its job
```
const { attributes } = afm;
```

**Comment by naibinh** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7122/ : SUCCESS in 6m 25s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3099/ : SUCCESS in 3s

**Comment by PhanDoThuyVi** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7125/ : SUCCESS in 3m 37s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3102/ : SUCCESS in 8s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3723/ : SUCCESS in 15m 21s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7126/ : SUCCESS in 3m 49s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1723/ : SUCCESS in 15s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/2530/ : SUCCESS in 20m 29s

**Pull request merged**